docs/development.md

```diff
@@ -0,0 +1,775 @@
+The CLI that is provided by this package will consume the invocation
+configuration and construct a state machine to execute work in a specific order.
+
+The `invocationValidator` function will run first to ensure the integration is
+capable of interfacing with a given provider. Next, the `getStepStartStates`
+will run to get a list of integration phases and steps that should be executed.
+Finally, the phases in `integrationStepPhases` are then executed in sequence.
+
+### What's in the `IntegrationExecutionContext` and `IntegrationExecutionStepContext`?
+
+The `IntegrationExecutionContext` contains some basic utilities to provide some
+information about the entity data.
+
+It includes a `logger` that can be used to display additional information
+throughout an integration run. An `instance` field containing the integration
+instance, which contains configuration field values stored under a `config`
+field.
+
+The `IntegrationExecutionStepContext` contains the same data stored in the
+`IntegrationExecutionContext` but also contains a `jobState` object that
+provides utilities for collecting and validating graph data. More details about
+the `jobState` is detailed in the [Data collection](# Data collection) section
+below.
+
+### Additional utilities
+
+#### Graph model validation
+
+To assist with keeping data consistent with our graph data model, the
+integration SDK will expose utility functions (`createIntegrationEntity` and
+`createIntegrationRelationship`) for validating entities and relationships based
+on their assigned `_class`.
+
+These functions will automatically validate that an entity contains required
+fields.
+
+#### Raw data collection
+
+In addition to performing validation, the `createIntegrationEntity` and
+`createIntegrationRelationship` will also automatically encode and store the
+original provider data under a `__rawData` field. For now, this will always
+assume that data coming in was stored as `json`, but support for other data
+types will come later.
+
+### Auto-magic behind the framework
+
+#### Logging
+
+When running an integration via the CLI tool will automatically log out
+information about the state changes that the integration. This allows for
+developers to keep track of what phase and step an integration is performing
+without the need to explicily add logging information themselves.
+
+#### Data collection
+
+The `executionContext` that is provided in the `executionHandler` step exposes a
+`jobState` utility that can be used to collect entity and relationship data via
+`addEntities` and `addRelationship` functions. The `jobState` utility will
+automatically flush the data to disk as a certain threshold of entities and
+relationships is met. The data flushed to disk are grouped in folders that based
+on the phase and step that was run. Flushed data will be replicated to a
+separate directory that groups data by an entity or relationship's `_type`.
```

Comment:
Will it actually be replicated, as in copied? If we're using the filesystem, can we just use links?

Reply:
My OG thought was that each step would produce a big dump of entities and relationships at the end of each step, but since there is already work being done to split up data via `_type`, there's no real reason why we couldn't just produce some smaller files and create links. Will update.